services/deduplicate/index.ts

```diff
@@ -0,0 +1,154 @@
+import { select, dispatch } from '@wordpress/data';
+
+const usedIds = new Map();
+const curatedIds = new Map();
+
+let running = false;
+let redo = false;
+
+interface Block {
+  attributes: {
+    backfillPosts?: number[];
+    numberOfPosts?: number;
+    posts?: number[];
+    postTypes?: string[];
+    query?: {
+      include?: number[];
+    }
+  },
+  clientId: string;
+  name: string;
+  innerBlocks?: Block[];
+}
+
+/**
+ * Checks if a post has been used already on this page. If so, return false. If not
+ * add it to the list and return true.
+ *
+ * @param {number|string} id The post id to check.
+ * @returns boolean
+ */
+export function deduplicate(id: number | string): boolean {
+  if (!id) {
+    return true;
+  }
+  const idNumber = typeof id === 'string' ? parseInt(id, 10) : id;
+  if (usedIds.has(idNumber)) {
+    return false;
+  }
+  usedIds.set(idNumber, true);
+  return true;
+}
+
+/**
+ * Resets the list of used ids.
+ */
+export function resetUsedIds() {
+  usedIds.clear();
+  curatedIds.clear();
+}
+
+export default {
+  deduplicate,
+  resetUsedIds,
+};
+
+/**
+ * This is the main function to update all pinned posts. Call it whenever a pinned post
+ * changes or the query settings change.
+ */
+export function mainDedupe() {
```

Comment:
This function is big and contains a lot of logic. As a reader of the code some comments would be helpful and make the reading of the mainDedupe flow a bit smoother.